--- src/shell/ash.c	2021-08-14 16:12:25.654298058 +0200
+++ src.newB/shell/ash.c	2021-08-14 16:10:01.126573612 +0200
@@ -10809,13 +10809,15 @@ preadbuffer(void)
 static void
 nlprompt(void)
 {
-	g_parsefile->linno++;
+	if(!doingtrap)
+		g_parsefile->linno++;
 	setprompt_if(doprompt, 2);
 }
 static void
 nlnoprompt(void)
 {
-	g_parsefile->linno++;
+	if(!doingtrap)
+		g_parsefile->linno++;
 	needprompt = doprompt;
 }
 
@@ -12505,7 +12507,8 @@ checkend: {
 
 		if (c == '\n' || c == PEOF) {
 			c = PEOF;
-			g_parsefile->linno++;
+			if(!doingtrap)
+				g_parsefile->linno++;
 			needprompt = doprompt;
 		} else {
 			int len_here;
