--- src/shell/ash.c	2021-08-14 18:03:15.799456855 +0200
+++ src.newC/shell/ash.c	2021-08-14 17:46:19.127778637 +0200
@@ -2337,7 +2337,8 @@ lookupvar(const char *name)
 				fmtstr(linenovar+7, sizeof(linenovar)-7, "%d", lineno);
 			} else
 			if (v->var_text == funcnamevar) {
-				fmtstr(funcnamevar+9, sizeof(funcnamevar)-9, "%s", funcname);
+				fmtstr(funcnamevar+9, sizeof(funcnamevar)-9, "%s", 
+					funcname ? funcname : "");
 			}
 			return var_end(v->var_text);
 		}
