--- src/libbb/loop.c	2021-08-12 17:06:57.876646932 +0200
+++ src.new/libbb/loop.c	2021-08-12 17:02:42.800025588 +0200
@@ -81,9 +81,26 @@ int FAST_FUNC del_loop(const char *devic
 /* Obtain an unused loop device number */
 int FAST_FUNC get_free_loop(void)
 {
+	static short found = -1;
+
 	int fd;
 	int loopdevno;
-	return -2;
+
+	if(found == -1) {
+		char buf[32];
+		FILE *fp;
+		found = 0;
+		fp = fopen("/proc/filesystems", "r");
+		if(!fp) return -2;
+		while(fgets(buf, 32, fp)) {
+			if(strstr(buf, "devtmpfs")) {
+				found = 1;
+				break;
+			}
+		}
+		fclose(fp);
+	}
+	if(!found) return -2;
 
 	fd = open("/dev/loop-control", O_RDWR | O_CLOEXEC);
 	if (fd == -1)
