--- src/shell/ash_test/run-all	2021-01-01 11:52:27.000000000 +0100
+++ src.1/shell/ash_test/run-all	2021-09-07 09:26:05.124109288 +0200
@@ -48,7 +48,8 @@ do_test()
 			*)
 			echo -n "$1/$x:"
 			sh "$x" >"$TOPDIR/$noslash-$x.fail" 2>&1 && \
-			{ { echo " ok"; rm "$TOPDIR/$noslash-$x.fail"; } || echo " fail"; }
+				{ { echo " ok"; rm "$TOPDIR/$noslash-$x.fail"; } \
+					|| echo " fail"; }
 			;;
 		esac
 	done
@@ -61,11 +62,12 @@ do_test()
 #		echo Running test: "$x"
 		echo -n "$1/$x:"
 		{
-			"$THIS_SH" "./$x" >"$name.xx" 2>&1
+			"$THIS_SH" "./$x" 2>&1 | \
+				grep -ve "^ash: using fallback suid method$" >"$name.xx" 
 			diff -u "$name.xx" "$name.right" >"$TOPDIR/$noslash-$x.fail" \
-			&& rm -f "$name.xx" "$TOPDIR/$noslash-$x.fail"
+				&& rm -f "$name.xx" "$TOPDIR/$noslash-$x.fail"
 		} && echo " ok" || echo " fail"
-		done
+	done
 	)
 }
 
