diff --git a/tinycore/changes/afterboot.sh b/tinycore/changes/afterboot.sh
index 7da49e8..432fb98 100644
--- a/tinycore/changes/afterboot.sh
+++ b/tinycore/changes/afterboot.sh
@@ -81,12 +81,26 @@ if [ "$ntdir" == "" ]; then
 	
 fi
 
+infotime -n "Loading TCZ archives: "
 if [ -d $tcdir/tcz ]; then
-	infotime -n "Loading TCZ archives: "
-	su - tc -c "tce-load -i $tcdir/tcz/*.tcz" | \
-		grep -v "is already installed!" | \
-			tr \\n ' ' | grep . || \
-				echo "no extra tcz!"
+	cd $tcdir/tcz
+	if [ -e onboot.lst ]; then
+		list=$(cat onboot.lst 2>/dev/null | cut -d# -f1 | tr \\n ' ')
+	else
+		list='*.tcz'
+	fi
+	n=$(echo "$list" | wc -w)
+	if [[ $n -gt 0 ]]; then
+		su tc -c "tce-load -i $list" | \
+			grep -v "is already installed!" | \
+				tr \\n ' ' | grep . || \
+					echo "no extra tcz!"
+	else
+		echo "none listed"
+	fi
+	cd - >/dev/null
+else
+	echo "folder does not exist!"
 fi
 
 infotime "Mounting local drives in read only..." ##############################
diff --git a/tinycore/changes/tcz2tce.sh b/tinycore/changes/tcz2tce.sh
index b141e2e..c442393 100644
--- a/tinycore/changes/tcz2tce.sh
+++ b/tinycore/changes/tcz2tce.sh
@@ -22,7 +22,7 @@ cd $mntdir
 if [ "$1" == "back" -o "$1" == "reverse" ]; then
 	if [ -d tce ]; then
 		mkdir -p tcz
-		mv -f $optional/*.tcz tcz
+		mv -f $optional/*.tcz tce/onboot.lst tcz
 		if [ "$1" == "reverse" ]; then
 			rm -rf tce
 			mkdir -p /tmp/tce
@@ -38,10 +38,13 @@ if [ "$1" == "back" -o "$1" == "reverse" ]; then
 fi
 
 mkdir -p $optional
-ls -1 tcz/*.tcz 2>/dev/null && \
-	mv -f tcz/*.tcz $optional
+mv -f tcz/*.tcz $optional 2>/dev/null
+mv -f tcz/onboot.lst $mntdir/tce 2>/dev/null
+rm -rf tcz
 cd $optional
-ls -1 *.tcz >../onboot.lst 2>/dev/null
+if [ ! -e ../onboot.lst ]; then
+	ls -1 *.tcz >../onboot.lst 2>/dev/null
+fi
 rm -f /etc/sysconfig/tcedir
 ln -sf $mntdir/tce /etc/sysconfig/tcedir
 sync
