--- src/shell/ash.c	2021-08-22 00:50:59.274681664 +0200
+++ src.5/shell/ash.c	2021-08-22 00:49:52.458681278 +0200
@@ -9288,12 +9288,14 @@ evaltree(union node *n, int flags)
 				static bool recursive = 0;
 				if(!recursive) {
 					int savelineno = lineno;
+					savestatus = exitstatus;
 					doingtrap = 1;
 					recursive = 1;
 					evalstring(trap[NSIG], 0);
 					recursive = 0;
 					doingtrap = 0;
 					lineno = savelineno;
+					exitstatus = savestatus;
 					if(eflag)
 						exitshell();
 			}
